transformations.go

```diff
@@ -0,0 +1,238 @@
+	var sequenceBuffer strings.Builder
+	for _, feature := range annotatedSequence.Features {
+		if feature.Type == "CDS" {
+			sequenceBuffer.WriteString(feature.getSequence())
+		}
+	}
+	optimizationTable := codonTable.generateOptimizationTable(sequenceBuffer.String())
+	for _, aminoAcid := range aminoAcids {
+		codons.WriteString(optimizationTable[string(aminoAcid)].Pick().(string))
+	}
+	return codons.String()
+}
+
+// Generate map of amino acid -> codon chooser
+func (codonTable CodonTable) generateOptimizationTable(sequence string) map[string]weightedRand.Chooser {
+	rand.Seed(time.Now().UTC().UnixNano())
+
+	sequence = strings.ToUpper(sequence)
+	var optimizationMap = make(map[string]weightedRand.Chooser)
+	codonFrequencyMap := GetCodonFrequency(sequence)
+
+	for _, aminoAcid := range codonTable.AminoAcids {
+		// Get list of triplets and their weights
+		codonChoices := make([]weightedRand.Choice, len(aminoAcid.Codons))
+
+		// Get sum of codon occurences for particular amino acid
+		codonOccurenceSum := 0
+		for _, codon := range aminoAcid.Codons {
+			triplet := codonFrequencyMap[codon.Triplet]
+			codonOccurenceSum += triplet
+		}
+
+		// Threshold codons that occur less than 10% for coding a particular amino acid
+		for _, codon := range aminoAcid.Codons {
+			if float64(codonFrequencyMap[codon.Triplet])/float64(codonOccurenceSum) < 0.10 {
+				codonFrequencyMap[codon.Triplet] = 0
+			}
+		}
+
+		// apply thresholded weights
+		for _, codon := range aminoAcid.Codons {
+			codonChoices = append(codonChoices, weightedRand.Choice{Item: codon.Triplet, Weight: uint(codonFrequencyMap[codon.Triplet])})
+		}
+
+		optimizationMap[aminoAcid.Letter] = weightedRand.NewChooser(codonChoices...)
+
+	}
+	return optimizationMap
+}
+
+// GetCodonFrequency takes a DNA sequence and returns a hashmap of its codons and their frequencies.
+func GetCodonFrequency(sequence string) map[string]int {
+
+	codonFrequencyHashMap := map[string]int{}
+	var currentCodon strings.Builder
+
+	for _, letter := range sequence {
+
+		// add current nucleotide to currentCodon
+		currentCodon.WriteRune(letter)
+
+		// if current nucleotide is the third in a codon add to hashmap
+		if currentCodon.Len() == 3 {
+			// if codon is already initalized in map increment
+			if _, ok := codonFrequencyHashMap[currentCodon.String()]; ok {
+				codonFrequencyHashMap[currentCodon.String()]++
+				// if codon is not already initalized in map initialize with value at 1
+			} else {
+				codonFrequencyHashMap[currentCodon.String()] = 1
+			}
+			// reset codon string builder for next codon.
+			currentCodon.Reset()
+		}
+	}
+	return codonFrequencyHashMap
+}
+
+// Generate map of codons -> amino acid
+func (codonTable CodonTable) generateTranslationTable() map[string]string {
+	var translationMap = make(map[string]string)
+	for _, aminoAcid := range codonTable.AminoAcids {
+		for _, codon := range aminoAcid.Codons {
+			translationMap[codon.Triplet] = aminoAcid.Letter
+		}
```

Comment:
Maybe I don't know go very well. But is this a `for loop` in a `for loop` in a `map` >.>